lighthouse-core/test/gather/computed/main-resource-test.js

```diff
@@ -0,0 +1,51 @@
+/**
+ * @license Copyright 2017 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/* eslint-env mocha */
+
+const Runner = require('../../../runner.js');
+const assert = require('assert');
+
+describe('MainResource computed artifact', () => {
+  let computedArtifacts;
+
+  beforeEach(() => {
+    computedArtifacts = Runner.instantiateComputedArtifacts();
+  });
+
+  it('returns an artifact', () => {
+    const record = {
+      statusCode: 404,
+    };
+    const networkRecords = [
+      record,
+    ];
+
+    return computedArtifacts.requestMainResource(networkRecords).then(output => {
+      assert.equal(output, record);
+    });
+  });
+
+  it('should ignore redirects', () => {
+    const record = {
+      statusCode: 404,
+    };
+    const networkRecords = [
+      {
+        statusCode: 301,
+      },
+      {
+        statusCode: 302,
+      },
+      record,
+    ];
+
+    return computedArtifacts.requestMainResource(networkRecords).then(output => {
+      assert.equal(output, record);
+    });
+  });
+});
```

Comment:
note we also have some devtoolsLogs and networkRecords in `lighthouse-core/test/fixtures/`, e.g. one from a wikipedia redirect that might be good for a test too

Reply:
Hm… if we use that fixture, we will be also testing NetworkRecords artefact, so it won't be a fully independent unit test. Also, wikipedia fixture does a bunch of redirects, just like the test above - so, IMO we will be repeating the same test. WDYT?

Reply:
> Hm… if we use that fixture, we will be also testing NetworkRecords artifact, so it won't be a fully independent unit test. Also, wikipedia fixture does a bunch of redirects, just like the test above - so, IMO we will be repeating the same test. WDYT?

Wellllll, it depends on if you buy that Lighthouse tests are really unit tests. Testing philosophies aside, our unit tests haven't been pure since the beginning, and are really a mixture of unit and integration/functional tests, so this won't be ruining anything. :)

This mixture is especially useful for something as complicated as talking to Chrome over the debugger protocol, where any mock we write is in danger of simplifying too much (or, conversely, making realistic mocks too much work to write).

Having the tests of the idealized network records above and the unit tests of devtoolsLogs -> networkRecords (in `network-recorder-test.js`) means that we should be able to isolate future issues in each of those places quickly, while using real networkRecords/devtoolsLogs here means that we can do a test of these two or three units in concert much more cheaply and quickly than firing up a smokehouse test for each case.

And if we change the heuristic for finding the main resource, the unit tests will need to change but a still working `wikipedia-redirect.devtoolslog.json` test (and maybe the h2 push fixture?) will be a good indicator that the change was successful.

Everyone has different feelings on where these testing borders should be, and maybe we should have set up separate unit vs functional tests from the beginning, but here we are :) WDYT?

Reply:
This absolutely makes sense, I just failed to realize that LH unit tests are a mix of different types of tests. I've added a test based on the wikipedia fixture.